### PR TITLE
Design panel: Text input, Media picker and Link panel

### PR DESCRIPTION
**Pull request by wassgha**
Partial for #253 and #261 

### Changes
- Implements a Text input component for usage in CTA link URLs, alt tags, descriptions, etc.
- Implements a Media input component for usage in video posters and CTA links (doubles as both circular and rectangular forms) Screen Shot 2020-03-10 at 1 20 16 AM
- Adds ability to change link descriptions and brand icons Screen Shot 2020-03-10 at 1 44 56 AM
- Styles the links panel correctly
- Adds ability to add links to video components (used to be broken)

### CODE REVIEW

File: assets/src/edit-story/components/panels/link.js [hunk 1]
```diff
@@ -31,83 +31,66 @@ import { __ } from '@wordpress/i18n';
...
+      setState((originalState) => ({
+        ...originalState,
+        [property]: value,
+      })),
+    [setState]
+  );
+  const handleChangeIcon = useCallback(
+    (image) => {
+      const icon =
+        image.sizes && image.sizes.medium ? image.sizes.medium.url : image.url;
```
Comment (obetomuniz) [inline assets/src/edit-story/components/panels/link.js]: I think that `image?.sizes?.medium ? ... : ....` will work in the same way.

Comment (wassgha) [reply assets/src/edit-story/components/panels/link.js]: Yup

File: assets/src/edit-story/utils/whyDidYouUpdate.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
+/*
```
Comment (swissspidy) [inline assets/src/edit-story/utils/whyDidYouUpdate.js]: Please remove this file in favor of `useWhyDidYouUpdate.js`, as per #506

**Review by swissspidy — APPROVED** [review]

File: assets/src/edit-story/components/form/inputGroup.js [hunk 3]
```diff
@@ -14,6 +14,8 @@
  * limitations under the License.
  */
 
+// !LEGACY: DO NOT USE.
```
Comment (dvoytenko) [inline assets/src/edit-story/components/form/inputGroup.js:17]: That will work :)

File: assets/src/edit-story/components/form/text.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+};
+
+TextInput.defaultProps = {
+  className: null,
+  disabled: false,
+  isMultiple: false,
+  flexBasis: 100,
+  textCenter: false,
+  clear: false,
+  ariaLabel: __('Standard input', 'web-stories'),
```
Comment (dvoytenko) [inline assets/src/edit-story/components/form/text.js:160]: IMHO this is not a good default label.

File: assets/src/edit-story/components/panels/link.js [hunk 5]
```diff
@@ -116,16 +98,19 @@ function LinkPanel({ selectedElements, onSetProperties }) {
...
-    if (state.link?.url) {
-      populateMetadata(state.link?.url);
+    if (state?.url) {
+      populateMetadata(state?.url);
+    } else if (state.url === '' || state.desc || state.icon) {
+      setState({ url: null, desc: null, icon: null });
+      onSetProperties({ link: null });
     }
-  }, [populateMetadata, state]);
+  }, [onSetProperties, populateMetadata, state]);
```
Comment (dvoytenko) [inline assets/src/edit-story/components/panels/link.js:107]: Ok, I'm getting rather stressed about the `useEffect(() => {}, [..., state])` pattern. That's what caused the infinite loop last time, no?

**Pull request merged**